File: scytale/ciphers/otp.py
from scytale.ciphers.base import Cipher


class OneTimePad(Cipher):
    name = "OneTimePad"

    def __init__(self, key=None):
        self.key = self.validate(key)

    def validate(self, key):
        if key is None:
            key = "IGBQU"
        key = key.upper()
        key = self.clean(key)
        return key

    def generate_pad(self, length):
        pad = self.key
        while len(pad) < length:
            pad += self.key
        return pad[:length]

    def encrypt(self, plaintext):
        plaintext = self.clean(plaintext.upper())
        pad = self.generate_pad(len(plaintext))
        a = self.alphabet
        return ''.join([
            a[(a.index(d) + a.index(p)) % len(a)]
            for d, p in zip(plaintext, pad)])

    def decrypt(self, ciphertext):
        pad = self.generate_pad(len(ciphertext))
        a = self.alphabet
        return ''.join([
            a[(a.index(d) - a.index(p)) % len(a)]
            for d, p in zip(ciphertext, pad)])

File: scytale/static/js/app/otp.js
(function($, APP) {

    function encrypt(plain, pad, alphabet) {
        var cipher = $.map(plain.split(""), function(pl, i) {
            var pa = pad[i % pad.length];
            pl = alphabet.indexOf(pl);
            pa = alphabet.indexOf(pa);
            var c = (pl + pa) % alphabet.length;
            return alphabet[c];
        });
        return cipher.join("");
    }

    function decrypt(cipher, pad, alphabet) {
        var plain = $.map(cipher.split(""), function(c, i) {
            var pa = pad[i % pad.length];
            c = alphabet.indexOf(c);
            pa = alphabet.indexOf(pa);
            // Need to add alphabet.length in order to force result to be
            // positive.
            var pl = (c - pa + alphabet.length) % alphabet.length;
            return alphabet[pl];
        });
        return plain.join("");
    }

    function clean(elm, alphabet) {
        var val = elm.val().toUpperCase(),
            not_alphabet = new RegExp("[^" + alphabet + "]", "g");
        val = val.replace(not_alphabet, "");
        elm.val(val);
    }

    APP.otp = function(alphabet) {
        var plain = $("#plaintext");
        var pad = $("#pad");
        var cipher = $("#ciphertext");

        plain.on("change input", function() {
            clean(plain, alphabet);
            cipher.val(encrypt(plain.val(), pad.val(), alphabet));
        });

        pad.on("change input", function() {
            clean(pad, alphabet);
            if (pad.val().length > 0) {
                cipher.val(encrypt(plain.val(), pad.val(), alphabet));
            }
        });

        cipher.on("change input", function() {
            clean(cipher, alphabet);
            plain.val(decrypt(cipher.val(), pad.val(), alphabet));
        });
    };
})(jQuery, APP || {});
